kamu-adapter-http/src/ws_smart_transfer_protocol_axum_server.rs

```diff
@@ -318,11 +350,251 @@ pub async fn dataset_pull_ws_handler(
+                },
+            )
+            .await;
+
+            if let Err(e) = response_result_objects {
+                Err(SmartProtocolPushServerError::PushObjectResponseWriteFailed(
+                    e,
+                ))
+            } else {
+                Ok(request.is_truncated)
+            }
+        }
+        Err(e) => Err(SmartProtocolPushServerError::PushObjectRequestReadFailed(e)),
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+async fn try_handle_push_complete(
+    socket: &mut axum::extract::ws::WebSocket,
+    dataset_builder: &dyn DatasetBuilder,
+) -> Result<(), SmartProtocolPushServerError> {
+    let maybe_complete_request = read_payload::<DatasetPushComplete>(socket).await;
+
+    match maybe_complete_request {
+        Ok(_) => {
+            tracing::debug!("Push client sent a complete request. Commiting the dataset");
+
+            match dataset_builder.finish().await {
+                Ok(_) => {}
+                Err(e) => {
+                    tracing::debug!("Committing dataset failed with error: {}", e);
+                    return Err(SmartProtocolPushServerError::Internal(e.int_err()));
+                }
+            }
+
+            tracing::debug!("Sending complete confirmation");
+
+            let response_result_objects = write_payload::<DatasetPushCompleteConfirmed>(
+                socket,
+                DatasetPushCompleteConfirmed {},
+            )
+            .await;
+
+            if let Err(e) = response_result_objects {
+                Err(SmartProtocolPushServerError::PushCompleteResponseWriteFailed(e))
+            } else {
+                Ok(())
+            }
+        }
+        Err(e) => Err(SmartProtocolPushServerError::PushCompleteRequestReadFailed(
+            e,
+        )),
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+async fn discard_dataset_building_on_error(
+    dataset_builder: &dyn DatasetBuilder,
+    e: SmartProtocolPushServerError,
+) {
+    tracing::debug!("Push process aborted with error: {}", e);
+    match dataset_builder.discard().await {
+        Ok(_) => {}
+        Err(e) => {
+            tracing::debug!("Discard dataset build error: {}", e);
+        }
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
 pub async fn dataset_push_ws_handler(
-    mut _socket: axum::extract::ws::WebSocket,
-    _dataset: Arc<dyn Dataset>,
+    mut socket: axum::extract::ws::WebSocket,
+    dataset_builder: Arc<Box<dyn DatasetBuilder>>,
+    dataset_url: Url,
 ) {
-    unimplemented!("Push smart protocol flow to be implemented")
+    let dataset = dataset_builder.as_dataset();
+    let push_request = match handle_push_request_initiation(&mut socket, dataset).await {
+        Ok(push_request) => push_request,
+        Err(e) => {
+            discard_dataset_building_on_error(dataset_builder.as_ref().as_ref(), e).await;
```

Comment:
Can we have a wrapper function that calls `discard_dataset_building_on_error` based on whether inner function succeeds or fails. Inner function will be able to use `?` operator for error handling, and avoid repeating this discard call in every place an error may originate from.

Reply:
Done